### DISCUSSION TITLE
Can't use a DataTemplate from a ResourceDictionary via DataTemplate selector

### DISCUSSION THREAD

**Discussion by Sekoree** [AvaloniaUI/Avalonia#15547] [Q&A]
Hi, I haven't used ResourceDictionaries much in the past so this might be a newbie mistake, but I'm a stuck.

I have DataTemplates in a ResourceDictionary that I'd like to use for an Expanders HeaderTemplate based on a selector (that implements IDataTemplate)

The ResourceDictionary looks like this:
```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    <DataTemplate x:Key="AudioCacheItem" 
                  DataType="<DataType>">
        <!-- Content -->
    </DataTemplate>
</ResourceDictionary>
```

The Template selector is this:
(this was partially taken from [this](https://stackoverflow.com/a/63207944) StackOverflow answer)
```
public class HeaderTemplateSelector : IDataTemplate
{
    [Content]
    public Dictionary<string, IDataTemplate> Templates { get; } = new();

    public Control? Build(object? param)
    {
        return param switch
        {
            AudioCacheItemViewModel 
                => Templates.TryGetValue("AudioCacheItem", out var template)
                ? template.Build(param)
                : null,
            _ => null
        };
    }

    public bool Match(object? data)
    {
        //some matching code
    }
}
```

The selector in my controls Resources is set like this:
```
<components:HeaderTemplateSelector x:Key="HeaderTemplateSelector">
    <DataTemplate x:Key="AudioCacheItem"
                  Content="{StaticResource AudioCacheItem}" />
</components:HeaderTemplateSelector>
```

This is set as a StaticResource to the Expander HeaderTemplate. 

There are no errors prior to building, but the build failed with the following message and I can't quite make sense of it.
``Error AVLN2000 Avalonia: Internal compiler error: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index') (ResolveContentPropertyTransformer) Line 20, position 14.
``

When I add a public setter to the ``Templates`` property the Error changes to:
``Error AVLN2000 Avalonia: Internal compiler error: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index') (ResolveContentPropertyTransformer) Line 20, position 14.``

That position is selector in the Resources.

**Comment by timunie** [comment]
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/IDataTemplateSample

**Reply by Sekoree** [reply]
Thank you! 
It seems like I cant have different DataTemplates in different files then (which was my initial plan).
Maybe I'm also just overcomplcating things now instead of just making these into seperate controls and then adding thhose to the selector.

**Reply by timunie** [reply]
You can do. Just need to use App.Current.TryFindResource here. https://docs.avaloniaui.net/docs/guides/styles-and-resources/resources#consuming-resources-from-code
